azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+		ginutil.WriteInvalidParamError(c, err, "user",
+			"Unable to import when both user and token are omitted.")
+		return wharfapi.Token{}, err
 	}
 
-	var projectBranches []azureDevOpsBranch
-	for _, ref := range projectRefs.Value {
-		name := strings.TrimPrefix(ref.Name, "refs/heads/")
-		projectBranches = append(projectBranches, azureDevOpsBranch{
-			Name: name,
-			Ref:  ref.Name,
-		})
+	var token wharfapi.Token
+	if i.TokenID != 0 {
+		token, err = getTokenByIDWritesError(c, client, i.TokenID)
+		if err != nil {
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Unable to get token by id %v.", i.TokenID))
+			return wharfapi.Token{}, err
+		}
+	} else {
+		token, err = client.GetToken(i.Token, i.User)
+		if err != nil || token.TokenID == 0 {
+			token, err = client.PostToken(wharfapi.Token{Token: i.Token, UserName: i.User})
+			if err != nil {
+				fmt.Println("Unable to post token: ", err)
+				ginutil.WriteAPIClientWriteError(c, err,"Unable to get existing token or create new token.")
+				return wharfapi.Token{}, err
+			}
+		}
 	}
 
-	return projectBranches, nil
+	i.User = token.UserName
+	i.Token = token.Token
+	i.TokenID = token.TokenID
+	return token, nil
 }
 
-func getRepositories(i importBody, project string) ([]byte, error) {
-	urlPath, err := url.Parse(i.URL)
+// getOrPostProviderWritesError Writes an error to the gin Context when an error occurs!
+func getOrPostProviderWritesError(c *gin.Context, client wharfapi.Client,
+	token wharfapi.Token, i importBody) (wharfapi.Provider, error) {
+	var provider wharfapi.Provider
+	if i.ProviderID != 0 {
+		provider, err := client.GetProviderById(i.ProviderID)
+		if err != nil || provider.ProviderID == 0 {
+			fmt.Printf("Unable to get provider. %+v", err)
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Unable to get provider by id %v", i.ProviderID))
+			return wharfapi.Provider{}, err
+		}
+	} else {
+		provider, err := client.GetProvider(apiProviderName, i.URL, i.UploadURL, token.TokenID)
+		if err != nil || provider.ProviderID == 0 {
+			provider, err = client.PostProvider(
+				wharfapi.Provider{Name: apiProviderName, URL: i.URL, TokenID: token.TokenID})
+			if err != nil {
+				fmt.Println("Unable to post provider: ", err)
+				ginutil.WriteAPIClientWriteError(c, err,
+					fmt.Sprintf("Unable to get or create provider from '%s'.", i.URL))
+				return wharfapi.Provider{}, err
+			}
+		}
+	}
+
+	i.URL = provider.URL
+	return provider, nil
+}
+
+func getRepositoriesWritesError(c *gin.Context, i importBody, project azureDevOpsProject) (azureDevOpsRepositoryResponse, error) {
+	const apiVersion string = "5.0"
+
+	urlPath, err := requests.ConstructGetURL(i.URL, map[string][]string{
+		"api-version": {apiVersion},
+	}, "%v/%v/%v", i.Group, project.Name, apiRepositories)
 	if err != nil {
 		fmt.Println("Unable to get url: ", err)
-		return []byte{}, err
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
```

Comment:
nit: Use the JSON names for parameter errors

```suggestion
		ginutil.WriteInvalidParamError(c, err, "url", fmt.Sprintf("Unable to parse URL %q.", i.URL))
```